crates/ibc/src/applications/transfer/relay/send_transfer.rs

```diff
@@ -1,33 +1,109 @@
-use crate::applications::transfer::context::TokenTransferContext;
+use crate::applications::transfer::context::{
+    TokenTransferExecutionContext, TokenTransferValidationContext,
+};
 use crate::applications::transfer::error::TokenTransferError;
 use crate::applications::transfer::events::TransferEvent;
 use crate::applications::transfer::msgs::transfer::MsgTransfer;
 use crate::applications::transfer::packet::PacketData;
 use crate::applications::transfer::{is_sender_chain_source, Coin, PrefixedCoin};
-use crate::core::ics04_channel::handler::send_packet::send_packet;
+use crate::core::ics04_channel::handler::send_packet::{send_packet_execute, send_packet_validate};
 use crate::core::ics04_channel::packet::Packet;
 use crate::core::ics24_host::path::{ChannelEndPath, SeqSendPath};
 use crate::events::ModuleEvent;
-use crate::handler::{HandlerOutput, HandlerOutputBuilder};
 use crate::prelude::*;
 
-/// This function handles the transfer sending logic.
-/// If this method returns an error, the runtime is expected to rollback all state modifications to
-/// the `Ctx` caused by all messages from the transaction that this `msg` is a part of.
-pub fn send_transfer<Ctx, C>(
-    ctx: &mut Ctx,
-    output: &mut HandlerOutputBuilder<()>,
+pub fn send_transfer<C, Ctx>(ctx_a: &mut Ctx, msg: MsgTransfer<C>) -> Result<(), TokenTransferError>
+where
+    C: TryInto<PrefixedCoin> + Clone,
+    Ctx: TokenTransferExecutionContext,
+{
+    send_transfer_validate(ctx_a, msg.clone())?;
+    send_transfer_execute(ctx_a, msg)
+}
+
+pub fn send_transfer_validate<C, Ctx>(
+    ctx_a: &Ctx,
```

Comment:
A side opinion about our naming convention (my experience so far)

While using the convention for message fields has been super helpful, I think we can ease up on naming contexts. (in particular we have *_on_a and *_of_a , not *_a ) we only implement one context for each host and always interact with that one context. Keeping ctx instead of ctx_a and ctx_b would be just as simple, effective, without adding unnecessary complexity or suffixes. Yet, wherever there could be confusion, we still see our naming convention on a message related field or variable.